In [16]:
from math import sqrt
from scipy import stats
from pydataset import data

%matplotlib inline
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
# import viz # curriculum visualization examples
from math import sqrt

# Hypothesis Testing

For each of the following questions, formulate a null and alternative hypothesis (be as specific as you can be), then give an example of what a true positive, true negative, type I and type II errors would look like. Note that some of the questions are intentionally phrased in a vague way. It is your job to reword these as more precise questions that could be tested.

## Has the network latency gone up since we switched internet service providers?

Many people have likely heard the term latency being used before but what is latency exactly? In terms of network latency, this can be defined by the time it takes for a request to travel from the sender to the receiver and for the receiver to process that request. In other words, the round trip time from the browser to the server. It is obviously desired for this time to remain as close to 0 as possible, however, there can be a few things at play preventing your website latency times to remain low.

data sources: 
* Measure time it take to send data
* Measure time it takes website to load
* Measure the Rount Trip Time (RTT)
* Measure the Time to First Byte (TTFB)
    * $H_0$: The TTFB is the same since we switched internet service providers
    * $H_a$: The TTFB is lower (closer to 0) since we switched internt service providers
    
- True Negative:
    - high p- value
    - fail to reject $H_0$
    - old provider TTFB == new provider TTFB
- False Negative:
    - high p-value
    - fail to reject $H_0$
    - old provider TTFB == new provider TTFB
    - but really we might be pinging only some of the routers, causing us to get false information
- True Positive:
    - low p-value
    - reject $H_0$
    - new provider TTFB == 0.5
    - old provider TTFB == 1.0
- False Positive:
    - low p-value
    - reject $H_0$
    - new provider TTFB == 0.5
    - old provider TTFB == 1.0
    - we might be looking at a small sample, and when we look at the avg over multiple trials it actually turns out there is no difference.

## Is the website redesign any good?

data sources:
* customer servey
* SEO optimization rankings
* google trends
* sales impact
* click-through rates / number of visitors
    * $H_0$: There has been no chance in the number of visitors since the website redesign 
    * $H_a$: There is a higher number of website vistors since the website redisgn
    
- True Negative: 
    * high p-value
    * we accept the $H_0$
    * website vistors before redisign == website vistors after redesign
- False Negative:
    * high p-value
    * accept the $H_0$
    * The data could include figures from other countries where the redesign hasn't taken place, and this is boosting the number of visitors prior to the redesign
    * We might not be allowing enough time marketing to annouce the new redesigns
- True Positive:
    * Low p-value
    * reject $H_0$
    * vistors after website redesign -> 150,000
    * visitors before website redesign -> 100,000
- False positive: 
    * low p-value
    * reject $H_0$
    * vistors after website redesign -> 150,000
    * visitors before website redesign -> 100,000
    * Website visits could be boosted by a sales promotions, and once the promotion ends, visits decrease 
    



## Is our television ad driving more sales?

data sources:
* sales before and after ad
* google trends
* store footfall
* website visitors
* recall survey
    * $H_0$: There is no difference in survey responses between customers that recall the ad, versus customers that didn't recall the ad
    * $H_a$: Customers that recall the ad were more likely to say they would buy the product
    
- True Negative: 
    * high p-value
    * fail to reject $H_0$
    * \# of people willing to buy our product and recall ad == \# of people willing to buy our product and don't recall ad
    
- False Negative:
    * high p-value
    * fail to reject $H_0$
    * There could be observer bias, where customers are equally likely to say they will buy the product, but this might not necesarly reflect their actual behavior
    
- True positive:
    * low p-value
    * reject $H_0$
    * \# of customers willing to buy our product and recall ad == 20
    * \# of customers willing to buy our product and don't recall ad == 10
    
- False positive:
    * low p-value
    * reject $H_0$
    * undercoverage biase - we might have only sample a portion of the population that that watches TV, and we are not considering what customers that don't watch TV are thinking

# Exercise 2: T - Test

## Question 1:

Ace Realty wants to determine whether the average time it takes to sell homes is different for its two offices. A sample of 40 sales from office #1 revealed a mean of 90 days and a standard deviation of 15 days. A sample of 50 sales from office #2 revealed a mean of 100 days and a standard deviation of 20 days. Use a .05 level of significance.

* $H_0$: There is no difference in the average of time it takes to sell homes between the two offices in Ace Realty
* $H_a$: There is a difference in the average of time it takes to sell homes between the two offices

Data sources: 
   * Internal sales data

In [9]:
# Office 1:

xbar1 = 90
s1 = 15
n1 = 50

# office 2:

xbar2 = 100
s2 = 20
n2 = 50

degf = (n1 + n2) - 2

s_p = sqrt(
    ((n1 - 1) * s1**2 + (n2 - 1) * s2**2)
    /
    (n1 + n2 - 2)
)

t = (xbar1 - xbar2) / (s_p * sqrt(1 / n1 + 1 / n2))

p = stats.t(degf).sf(t) * 2

t, p

(-2.82842712474619, 1.9943269821793772)

In [13]:
alpha = 0.05

print(f'''
Becasue p ({p:.6f}) > alpha ({alpha}), we fail to reject the null hypothesis, meaning that there is no significant difference in average time to sell a home between the two branches''')


Becasue p (1.994327) > alpha (0.05), we fail to reject the null hypothesis, meaning that there is no significant difference in average time to sell a home between the two branches


## Question 2.
Load the mpg dataset and use it to answer the following questions:

In [23]:
mpg = data('mpg')

In [24]:
mpg.head()

manufacturer model  displ  year  cyl       trans drv  cty  hwy fl    class
1         audi    a4    1.8  1999    4    auto(l5)   f   18   29  p  compact
2         audi    a4    1.8  1999    4  manual(m5)   f   21   29  p  compact
3         audi    a4    2.0  2008    4  manual(m6)   f   20   31  p  compact
4         audi    a4    2.0  2008    4    auto(av)   f   21   30  p  compact
5         audi    a4    2.8  1999    6    auto(l5)   f   16   26  p  compact

### Is there a difference in fuel-efficiency in cars from 2008 vs 1999?

* $H_0$: There is no difference in fuel efficiency between cars from 2008 vs 1999
* $H_a$: Cars from 2008 have a better fuel efficienct than cars from 1999

In [25]:
mpg["fuel_efficiency"] = (mpg["cty"] + mpg["hwy"]) / 2

In [36]:
x1 = mpg[mpg.year == 1999].fuel_efficiency
x2 = mpg[mpg.year == 2008].fuel_efficiency

stats.ttest_ind(x1, x2)

Ttest_indResult(statistic=0.21960177245940962, pvalue=0.8263744040323578)

Because the p value (0.826) > alpha (0.05), we fail to reject the null hypothesis. 

This means that there is no statistical difference between cars from 1999 and 2008

* Are compact cars more fuel-efficient than the average car?


* $H_0$: There is no significant difference in fuel-efficiency between compact cars and the average car
* $H_a$: Compact cars are on average more fuel- efficient than the average car

In [85]:
x = mpg[mpg["class"] == "compact"].fuel_efficiency
mu = mpg.fuel_efficiency.mean()

t_stat, p = stats.ttest_1samp(x, mu)

print(f't = {t_stat:.3}')
print(f'p = {p:.4}')

t = 7.9
p = 4.199e-10


In [60]:
x.mean(), mu

(24.21276595744681, 20.14957264957265)

The p value is below the alpha, which means we reject the null hypothesis that compact cars, on average, have the same fuel-efficiency as the average car. 

* Do manual cars get better gas mileage than automatic cars?

* $H_0$: There is no difference in gas mileage between manual and automatic cars
* $H_a$: There is a difference in gas mileage between manual and automatic cars

In [73]:
x1 = mpg[mpg.trans.str.contains("auto")].fuel_efficiency
x2 = mpg[mpg.trans.str.contains("manual")].fuel_efficiency

stats.ttest_ind(x1,x2)

Ttest_indResult(statistic=-4.593437735750014, pvalue=7.154374401145683e-06)

In [79]:
print(f"avg fuel effienct for automatic cars: {x1.mean():.3}, and manual: {x2.mean():.3}")

avg fuel effienct for automatic cars: 19.1, and manual: 22.2


Becasue the p value is smaller than that alpha, we can reject the null hypothesis that there is no difference between transmission and fuel efficiency. 

For this particular case, it seems that manual cars are significantly different, on average, than automatic cars when it comes to fuel efficiency.

What if we wanted to see if the manual cars, on average, had better fuel efficiency? 

In [87]:
x = mpg[mpg.trans.str.contains("manual")].fuel_efficiency
mu = mpg.fuel_efficiency.mean()

t_stat, p = stats.ttest_1samp(x,mu)

print(f"t = {t_stat:.3}")
print(f"p = {p:.3}")

t = 3.53
p = 0.000703


Because the p value is still lower than alpha, we can conclude that there is infact a significant difference in fuel efficiency between manual cars and the rest.